core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -443,9 +466,27 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorIdle(executorId: String): Unit = synchronized {
     if (executorIds.contains(executorId)) {
       if (!removeTimes.contains(executorId) && !executorsPendingToRemove.contains(executorId)) {
+
+        val hasCachedBlocks =
+          executorsWithCachedBlocks.contains(executorId) ||
+            executorEndpoints.get(executorId).exists(_.askWithRetry[Boolean](HasCachedBlocks))
+
+        if (hasCachedBlocks) executorsWithCachedBlocks += executorId
+
+        val now = clock.getTimeMillis()
+        val timeout = {
+          if (hasCachedBlocks) {
+            val newExpiry = now + cachedExecutorTimeoutS * 1000
+            if (newExpiry < 0) Long.MaxValue // Overflow
+            else newExpiry
+          } else {
+            now + executorIdleTimeoutS * 1000
+          }
+        }
+
+        removeTimes(executorId) = timeout
```

Comment:
This would change the behavior of non-cached executors (of course, this change is definitely valid, since the current code does not consider overflows at all).